compiler/rustc_mir/src/interpret/eval_context.rs

```diff
@@ -815,31 +815,18 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         // Usually we want to clean up (deallocate locals), but in a few rare cases we don't.
         // In that case, we return early. We also avoid validation in that case,
         // because this is CTFE and the final value will be thoroughly validated anyway.
-        let (cleanup, next_block) = match frame.return_to_block {
-            StackPopCleanup::Goto { ret, unwind } => (
+        let (cleanup, next_block) = match (frame.return_to_block, unwinding) {
+            (StackPopCleanup::Goto { ret, .. }, false) => (true, Some(ret)),
+            (StackPopCleanup::Goto { unwind, .. }, true) => (
                 true,
-                Some(if unwinding {
-                    let def_id = frame.body.source.def_id();
-                    match unwind {
-                        StackPopUnwind::Cleanup(unwind)
-                            // `fn_sig()` can't be used on closures, but closures always have
-                            // "rust-call" ABI, which always allows unwinding anyway.
-                            if self.tcx.is_closure(def_id) || self.fn_can_unwind(
-                                self.tcx.codegen_fn_attrs(def_id).flags,
-                                self.tcx.fn_sig(def_id).abi(),
-                            ) =>
-                        {
-                            unwind
-                        }
-                        _ => {
-                            throw_ub_format!("unwind past a frame that does not allow unwinding")
-                        }
+                Some(match unwind {
+                    StackPopUnwind::Cleanup(unwind) => unwind,
+                    StackPopUnwind::NotAllowed => {
+                        throw_ub_format!("unwind past a frame that does not allow unwinding")
```

Comment:
```suggestion
                        throw_ub_format!("unwinding past a frame that does not allow unwinding")
```
This is more consistent with some of our other messages.

Will it be clear from the message which function this is that we can't unwind past? (i.e. from the spans that are shown)

Reply:
> Will it be clear from the message which function this is that we can't unwind past? (i.e. from the spans that are shown)

I think so, according the Miri test I added: https://github.com/rust-lang/miri/compare/d866c1c52f48bf562720383455b75c257bb1ad92..626638fbfe2fff34648dda29a34d59db498a6e52#diff-523417889d59d37a42fe1b22a2bad6922d9e211bef625d0688f8cb82488cdb0f

Reply:
I can't really tell that from the test, could you show full output?

Reply:
Hmm... Will it be unclear if `unwind` is a function pointer?
```rust
#[no_mangle]
extern "C" fn unwind() {
    panic!();
}
fn main() {
    extern "C" {
        fn unwind();
    }
    unsafe { unwind() }
}
```
```
error: Undefined Behavior: unwinding past a frame that does not allow unwinding
 --> output.rs:9:14
  |
9 |     unsafe { unwind() }
  |              ^^^^^^^^ unwinding past a frame that does not allow unwinding
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
          
  = note: inside `main` at /home/hyd-dev/Playground/output.rs:9:14
  = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at $RUST_SRC/library/core/src/ops/function.rs:227:5
  = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at $RUST_SRC/library/std/src/sys_common/backtrace.rs:125:18
  = note: inside closure at $RUST_SRC/library/std/src/rt.rs:49:18
  = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at $RUST_SRC/library/core/src/ops/function.rs:259:13
  = note: inside `std::panicking::try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panicking.rs:401:40
  = note: inside `std::panicking::try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at $RUST_SRC/library/std/src/panicking.rs:365:19
  = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panic.rs:433:14
  = note: inside `std::rt::lang_start_internal` at $RUST_SRC/library/std/src/rt.rs:34:21
  = note: inside `std::rt::lang_start::<()>` at $RUST_SRC/library/std/src/rt.rs:48:5

error: aborting due to previous error

Reply:
Maybe it's a bit problematic for a function pointer?
```rust
#[no_mangle]
extern "C" fn unwind() {
    panic!();
}

fn get_fn_ptr_from_somewhere_else() -> unsafe extern "C" fn() {
    unwind
}

fn main() {
    let fn_ptr = get_fn_ptr_from_somewhere_else();
    unsafe { fn_ptr() }
}
```
```
error: Undefined Behavior: unwinding past a frame that does not allow unwinding
  --> output.rs:12:14
   |
12 |     unsafe { fn_ptr() }
   |              ^^^^^^^^ unwinding past a frame that does not allow unwinding
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
           
   = note: inside `main` at output.rs:12:14
   = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at $RUST_SRC/library/core/src/ops/function.rs:227:5
   = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at $RUST_SRC/library/std/src/sys_common/backtrace.rs:125:18
   = note: inside closure at $RUST_SRC/library/std/src/rt.rs:49:18
   = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at $RUST_SRC/library/core/src/ops/function.rs:259:13
   = note: inside `std::panicking::try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panicking.rs:401:40
   = note: inside `std::panicking::try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at $RUST_SRC/library/std/src/panicking.rs:365:19
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panic.rs:433:14
   = note: inside `std::rt::lang_start_internal` at $RUST_SRC/library/std/src/rt.rs:34:21
   = note: inside `std::rt::lang_start::<()>` at $RUST_SRC/library/std/src/rt.rs:48:5

error: aborting due to previous error
```

@rustbot label -S-waiting-on-author S-waiting-on-review

Reply:
If you move this logic to before the `pop` call, the stack trace should include the called function.

Reply:
> If you move this logic to before the `pop` call, the stack trace should include the called function.

:+1: The output looks like this:
```
error: Undefined Behavior: unwinding past a frame that does not allow unwinding
  --> output.rs:2:1
   |
2  | / extern "C" fn unwind() {
3  | |     panic!();
4  | | }
   | |_^ unwinding past a frame that does not allow unwinding
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
           
   = note: inside `unwind` at output.rs:2:1
note: inside `main` at output.rs:12:14
  --> output.rs:12:14
   |
12 |     unsafe { fn_ptr() }
   |              ^^^^^^^^
   = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at $RUST_SRC/library/core/src/ops/function.rs:227:5
   = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at $RUST_SRC/library/std/src/sys_common/backtrace.rs:125:18
   = note: inside closure at $RUST_SRC/library/std/src/rt.rs:49:18
   = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at $RUST_SRC/library/core/src/ops/function.rs:259:13
   = note: inside `std::panicking::try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panicking.rs:401:40
   = note: inside `std::panicking::try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at $RUST_SRC/library/std/src/panicking.rs:365:19
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panic.rs:433:14
   = note: inside `std::rt::lang_start_internal` at $RUST_SRC/library/std/src/rt.rs:34:21
   = note: inside `std::rt::lang_start::<()>` at $RUST_SRC/library/std/src/rt.rs:48:5

error: aborting due to previous error
```